File: chaoxing-desktop/src/types/course.ts
export type CourseType = "all" | "course" | "archive" | "other";
export type JobType = "video" | "document" | "read" | "live" | "workid";

/** 课程信息 */
export interface Course {
  id: string;
  courseId: string;
  clazzId: string;
  cpi: string;
  title: string;
  teacher: string;
  desc: string;
  info: string;
  roleid: string;
  courseType: CourseType;
  courseTypeLabel: string;
}

/** 任务点 */
export interface CourseJob {
  id: string;
  name: string;
  jobType: JobType;
  typeLabel: string;
  isCompleted: boolean;
}

/** 章节知识点 */
export interface ChapterPoint {
  id: string;
  title: string;
  jobCount: number;
  hasFinished: boolean;
  needUnlock: boolean;
}

/** 带任务的章节节点 */
export interface ChapterSelectionPoint extends ChapterPoint {
  jobs: CourseJob[];
}

/** 章节树结构 */
export interface ChapterTree {
  hasLocked: boolean;
  points: ChapterPoint[];
}

/** 课程内容选择树 */
export interface CourseSelectionTree {
  hasLocked: boolean;
  points: ChapterSelectionPoint[];
}

/** 扁平任务项 */
export interface FlatCourseJob extends CourseJob {
  pointId: string;
  pointTitle: string;
  hasFinished: boolean;
  needUnlock: boolean;
}

/** 学习范围中的章节选择 */
export interface CoursePointSelection {
  pointId: string;
  selectedJobIds: string[];
}

/** 单门课程的学习范围 */
export interface CourseLearningSelection {
  courseId: string;
  clazzId: string;
  cpi: string;
  title: string;
  selectedPoints: CoursePointSelection[];
}

File: chaoxing-desktop/src-tauri/src/models/course.rs
use serde::{Deserialize, Serialize};

use crate::models::job::JobType;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Course {
    pub id: String,
    pub course_id: String,
    pub clazz_id: String,
    pub cpi: String,
    pub title: String,
    pub teacher: String,
    pub desc: String,
    pub info: String,
    pub roleid: String,
    pub course_type: CourseType,
    pub course_type_label: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CourseFolder {
    pub id: String,
    pub rename: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum CourseType {
    Course,
    Archive,
    Other,
}

impl CourseType {
    pub fn label(&self) -> &'static str {
        match self {
            Self::Course => "课程",
            Self::Archive => "归档",
            Self::Other => "其他",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CourseJob {
    pub id: String,
    pub name: String,
    pub job_type: JobType,
    pub type_label: String,
    pub is_completed: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChapterSelectionPoint {
    pub id: String,
    pub title: String,
    pub job_count: u32,
    pub has_finished: bool,
    pub need_unlock: bool,
    pub jobs: Vec<CourseJob>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CourseSelectionTree {
    pub has_locked: bool,
    pub points: Vec<ChapterSelectionPoint>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CoursePointSelection {
    pub point_id: String,
    pub selected_job_ids: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CourseTaskSelection {
    pub course_id: String,
    pub clazz_id: String,
    pub cpi: String,
    pub title: String,
    pub selected_points: Vec<CoursePointSelection>,
}
